### PR TITLE
Support meta["val"] that is a dict, for triton kernels and for the partitioner

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132531
* __->__ #132466
* #132356

Internally there's a model that's using memory_budget with the partitioner, and using custom triton kernels. The partitioner fails when encountering the triton ops because they don't have `meta["val"]`. This PR adds `meta["val"]`  to these fx graph nodes and then adds handling for `meta["val"]` being a dict in the partitioner. 

Differential Revision: [D60627813](https://our.internmc.facebook.com/intern/diff/D60627813)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60627813).

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 1]
```diff
@@ -582,7 +582,10 @@ def trace_triton_kernel_wrapper(proxy_mode, func_overload, node_args):
         proxy_args,
         name=func_overload.__name__ + "_proxy",
     )
-    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+    ret = track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/_higher_order_ops/triton_kernel_wrap.py:585]: ah, thanks for the pointer! added it in track_tensor_tree

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60627813).

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -437,9 +439,8 @@ def get_constant(c: Optional[_NestedTensors], idx: int) -> Optional[_NestedTenso
             # which does not participate in const-prop)
             assert constant is None
 
-            # if isinstance(proxy, fx.Proxy):
-            #    # BUG? This is guaranteed to be a no-op
-            #    set_meta(proxy, e)
+            if isinstance(proxy, fx.Proxy):
+                set_meta(proxy, e)
```
[the remark this answers is not in the document]
Comment (aorenste) [reply torch/fx/experimental/proxy_tensor.py:443]: You can add it back if you think it's doing something useful - but it doesn't seem to do anything (I guess if `proxy.node.meta['val'] is not None` then it sets it to `None`).

[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/fx/experimental/proxy_tensor.py:443]: It fixes the test added in this PR

[the remark this answers is not in the document]
Comment (aorenste) [reply torch/fx/experimental/proxy_tensor.py:443]: Oh - you're adding a case for `dict` above - yeah - definitely add this back in.

File: test/functorch/test_ac.py [hunk 3]
```diff
@@ -1,11 +1,22 @@
...
+from torch.utils._triton import has_triton
+from torch.utils.flop_counter import FlopCounterMode, register_flop_formula
+
+
+if has_triton():
+    # note: if we only import triton in the test, the test fails:
+    # def relu_kernel_(inp_ptr, out_ptr, sz, BLOCK_SIZE: tl.constexpr):
+    # NameError('tl is not defined')
+    import triton
+    import triton.language as tl
```
Comment (zou3519) [inline test/functorch/test_ac.py:19]: I'm confused, this should work if you import it in the test?

**Review by zou3519 — APPROVED** [review]

File: test/functorch/test_ac.py (hunk 3 above)
Comment (davidberard98) [reply test/functorch/test_ac.py:19]: filed an issue here: https://github.com/pytorch/pytorch/issues/132534

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60627813).

**Pull request closed without merging**